README.md
# Operation Registry Plugin

The operation registry plugin is the interface into the Apollo Platform's **operation registry** and enables operation **safelisting**, which allows selective execution based on the operation. Safelisting eliminates the risk of unexpected operations that could cause downtime from being run against a graph.

In order to enable safelisting, follow the [step by step guide in the Apollo docs](https://www.apollographql.com/docs/studio/operation-registry/). These steps describe how to extract and upload operations defined within client applications to [Apollo Studio](https://studio.apollographql.com) using the Apollo CLI. Once operations have been registered, this plugin for Apollo Server fetches the manifest of these operations from [Apollo Studio](https://studio.apollographql.com) and forbids the execution of any operations that are not in that manifest.

### Usage

The following example shows basic usage of the plugin with Apollo Server. First, add the plugin to your project's `package.json`:

```bash
npm install @apollo/server-plugin-operation-registry
```

Then, ensure Apollo Server has access to an [API key](https://www.apollographql.com/docs/studio/operation-registry/#6-start-apollo-server-with-apollo-studio-enabled), for example as the `APOLLO_KEY` environment variable:

```bash
APOLLO_KEY=<API_KEY> npm start
```

Next, enable the plugin by adding it to the `plugins` parameter to the Apollo Server options:

```js
const server = new ApolloServer({
  typeDefs,
  resolvers,
  subscriptions: false,
  plugins: [
    require("@apollo/server-plugin-operation-registry")({
      forbidUnregisteredOperations: true
    })
  ]
});
```

<details><summary>With federation, the setup follows the same `plugins` configuration:</summary>

```js
const { ApolloServer } = require("apollo-server");
const { ApolloGateway } = require("@apollo/gateway");

const gateway = new ApolloGateway({
  serviceList: [ /* services */ ],
});

const server = new ApolloServer({
  gateway,
  subscriptions: false,
  plugins: [
    require("@apollo/server-plugin-operation-registry")({
      forbidUnregisteredOperations: true
    })
  ]
});

server.listen().then(({ url }) => {
  console.log(`🚀 Server ready at ${url}`);
});
```
</details>

#### Variant

Clients can register their operations to a specific variant, so the plugin contains the `graphVariant` field to specify which variant to pull operation manifests from.

```js
const server = new ApolloServer({
  plugins: [
    require("@apollo/server-plugin-operation-registry")({
      graphVariant: "production"
    })
  ]
});
```


### Metrics

The plugin will transmit metrics regarding unregistered operations which can be viewed within [Apollo Studio](https://studio.apollographql.com). The following example shows the unregistered operations sent by a particular client:

<p align="center">
  <img
    src="https://cl.ly/2a5b9c82287d/download/clients-page.png"
    alt="The clients page showing unregistered operations"
  />
</p>


CHANGELOG.md
# @apollo/server-plugin-operation-registry

## 4.0.1

### Patch Changes

- [#133](https://github.com/apollographql/server-plugin-operation-registry/pull/133) [`374fc29`](https://github.com/apollographql/server-plugin-operation-registry/commit/374fc2987af8a77333e657185011c11c95c58d9e) Thanks [@renovate](https://github.com/apps/renovate)! - Update utils packages to v2 (which only drop node v12 support)

## 4.0.0

### Major Changes

- [#131](https://github.com/apollographql/server-plugin-operation-registry/pull/131) [`2e6b87d`](https://github.com/apollographql/server-plugin-operation-registry/commit/2e6b87dd09255d9476e4f3efa10ef953f13b4514) Thanks [@trevor-scheer](https://github.com/trevor-scheer)! - Drop support for Node.js v12

## 3.5.6

### Patch Changes

- [#68](https://github.com/apollographql/server-plugin-operation-registry/pull/68) [`c585bf2`](https://github.com/apollographql/server-plugin-operation-registry/commit/c585bf20e3b39ff6d82c0afbea2a1af68199b825) Thanks [@trevor-scheer](https://github.com/trevor-scheer)! - Remove the apollo-graphql dependency in favor of equivalent apollo-utils deps

## 3.5.5

### Patch Changes

- [#36](https://github.com/apollographql/server-plugin-operation-registry/pull/36) [`792d52f`](https://github.com/apollographql/server-plugin-operation-registry/commit/792d52fb6d4a560e823d5d79181ca47a256f973d) Thanks [@trevor-scheer](https://github.com/trevor-scheer)! - Add tsconfig files to published output

## 3.5.4

### Patch Changes

- [#8](https://github.com/apollographql/server-plugin-operation-registry/pull/8) [`6c7b54b`](https://github.com/apollographql/server-plugin-operation-registry/commit/6c7b54b4f1b429a73083a9ec622a5f420c2f1665) Thanks [@rubencodes](https://github.com/rubencodes)! - Fix an order of operations error when resolving the manifest and storage secret base URLs

## 3.5.3

### Patch Changes

- [#6](https://github.com/apollographql/server-plugin-operation-registry/pull/6) [`bba040b`](https://github.com/apollographql/server-plugin-operation-registry/commit/bba040bdf73d4b1b4cce7982d1dba7a4aa4d4745) Thanks [@trevor-scheer](https://github.com/trevor-scheer)! - Re-release under a new package name

## 3.5.2

### Patch Changes

- [#1](https://github.com/apollographql/server-plugin-operation-registry/pull/1) [`f2521b4`](https://github.com/apollographql/server-plugin-operation-registry/commit/f2521b4c418572c76cb959d3c663b075e7707b82) Thanks [@joel-burton](https://github.com/joel-burton)! - initial publish of migrated package


src/__tests__/agent.test.ts
import nock from 'nock';
import type { InMemoryLRUCache } from 'apollo-server-caching';
import { resolve as urlResolve } from 'url';
import {
  defaultAgentOptions,
  genericServiceID,
  genericStorageSecret,
  nockStorageSecret,
  nockBase,
  nockGoodManifestsUnderStorageSecret,
  defaultStore,
  defaultTestAgentPollSeconds,
  nockStorageSecretOperationManifest,
  genericApiKeyHash,
} from './helpers.test-helpers';
import Agent, { AgentOptions } from '../agent';
import type { Operation } from '../ApolloServerPluginOperationRegistry';
import {
  fakeTestBaseUrl,
  getStoreKey,
  getOperationManifestUrl,
  urlOperationManifestBase,
} from '../common';
import type { Logger } from 'apollo-server-types';

// These get a bit verbose within the tests below, so we use this as a
// sample store to pick and grab from.
const sampleManifestRecords: Record<string, Operation> = {
  a: {
    signature:
      'ba4573fca2e1491fd54b9f398490531ad6327b00610f2c1216dc8c9c4cfd2993',
    document:
      'mutation toggleMovieLike($id:ID!){toggleLike(id:$id){__typename id isLiked}}',
  },
  b: {
    signature:
      '32a21510374c3c9ad25e06424085c45ccde29bdbdedf8fa806c2bc6a2ffcdf56',
    document: '{nooks:books{author}}',
  },
  c: {
    signature:
      'c60ac6dfe19ba70dd9d6a29a275fae56296dcbb636eeaab55c3d9b7287c40a47',
    document: '{nooks:books{__typename author}}',
  },
};

const manifest = (...operations: Operation[]) => ({
  version: 2,
  operations,
});

describe('Agent', () => {
  describe('Basic', () => {
    const Agent = require('../agent').default;
    it('instantiates with proper options', () => {
      expect(new Agent({ ...defaultAgentOptions })).toBeInstanceOf(Agent);
    });

    it('instantiates with debug enabled', () => {
      expect(new Agent({ ...defaultAgentOptions, debug: true })).toBeInstanceOf(
        Agent,
      );
    });
  });

  describe('with manifest', () => {
    const forCleanup: {
      store?: InMemoryLRUCache;
      agent?: import('../agent').default;
    }[] = [];

    function createAgent({ ...args }: Partial<AgentOptions> = {}) {
      const options = { ...defaultAgentOptions, ...args };

      // We never actually let the Agent construct its own default store
      // since we need to pluck the store out to instrument it with spies.
      const store = options.store;
      const agent = new Agent(options);

      // Save all agents and stores we've created so we can properly
      // stop them and clean them up.
      forCleanup.push({ agent, store });
      return agent;
    }

    afterEach(() => {
      if (!nock.isDone()) {
        throw new Error(
          `Not all nock interceptors were used! Pending mocks: ${nock.pendingMocks()}`,
        );
      }

      let toCleanup;
      // Loop through the `forCleanup` constant and empty it out by popping
      // individual elements off the end and running the appropriate cleanup.
      while ((toCleanup = forCleanup.pop())) {
        if (toCleanup.agent) {
          toCleanup.agent.stop();
        }
      }
    });

    describe('manifest checking and store populating', () => {
      function expectStoreSpyOperationEach(
        spy: jest.SpyInstance,
        letters: string[],
      ) {
        letters.forEach((letter) => {
          const { signature, document } = sampleManifestRecords[letter];
          expect(spy).toHaveBeenCalledWith(getStoreKey(signature), document);
        });
      }

      async function expectStoreHasOperationEach(
        store: InMemoryLRUCache,
        letters: string[],
      ) {
        for (const letter of letters) {
          const { signature, document } = sampleManifestRecords[letter];
          await expect(
            store.get(getStoreKey(signature)),
          ).resolves.toStrictEqual(document);
        }
      }

      it('logs debug updates to the manifest on startup', async () => {
        nockStorageSecret(genericServiceID, genericApiKeyHash);
        nockGoodManifestsUnderStorageSecret(
          genericServiceID,
          genericStorageSecret,
          [
            sampleManifestRecords.a,
            sampleManifestRecords.b,
            sampleManifestRecords.c,
          ],
        );
        const relevantLogs: any = [];
        const logger: Logger = {
          debug: jest.fn().mockImplementation((...args: any[]) => {
            if (
              typeof args[0] === 'string' &&
              (args[0].match(/Checking for manifest changes/) ||
                args[0].match(/Incoming manifest ADDs/))
            ) {
              relevantLogs.push(args);
            }
          }),
          warn: jest.fn().mockImplementation((...args: any[]) => {
            if (
              typeof args[0] === 'string' &&
              (args[0].match(/Checking for manifest changes/) ||
                args[0].match(/Incoming manifest ADDs/))
            ) {
              relevantLogs.push(args);
            }
          }),
          info: () => {},
          error: () => {},
        };
        await createAgent({ logger }).start();

        expect(relevantLogs[0][0]).toBe(
          `Checking for manifest changes at ${urlResolve(
            fakeTestBaseUrl,
            getOperationManifestUrl(
              genericServiceID,
              genericStorageSecret,
              'current',
            ),
          )}`,
        );

        // Console should indicate the records have been added in order.
        expect(relevantLogs[1][0]).toBe(
          `Incoming manifest ADDs: ${sampleManifestRecords.a.signature}`,
        );
        expect(relevantLogs[2][0]).toBe(
          `Incoming manifest ADDs: ${sampleManifestRecords.b.signature}`,
        );
        expect(relevantLogs[3][0]).toBe(
          `Incoming manifest ADDs: ${sampleManifestRecords.c.signature}`,
        );

        expect(relevantLogs.length).toBe(4);
      });

      it('populates the manifest store after starting', async () => {
        nockStorageSecret(genericServiceID, genericApiKeyHash);
        nockGoodManifestsUnderStorageSecret(
          genericServiceID,
          genericStorageSecret,
          [
            sampleManifestRecords.a,
            sampleManifestRecords.b,
            sampleManifestRecords.c,
          ],
        );
        const store = defaultStore();
        const storeSetSpy = jest.spyOn(store, 'set');
        await createAgent({ store }).start();

        // There are three operations in the manifest above.
        expect(storeSetSpy).toHaveBeenCalledTimes(3);
        expectStoreSpyOperationEach(storeSetSpy, ['a', 'b', 'c']);
        await expectStoreHasOperationEach(store, ['a', 'b', 'c']);

        storeSetSpy.mockRestore();
      });

      it('starts polling successfully', async () => {
        nockStorageSecret(genericServiceID, genericApiKeyHash);
        nockGoodManifestsUnderStorageSecret(
          genericServiceID,
          genericStorageSecret,
          [
            sampleManifestRecords.a,
            sampleManifestRecords.b,
            sampleManifestRecords.c,
          ],
        );
        const store = defaultStore();
        const storeSetSpy = jest.spyOn(store, 'set');
        const storeDeleteSpy = jest.spyOn(store, 'delete');
        const agent = createAgent({ store });
        // @ts-ignore
        jest.useFakeTimers({
          legacyFakeTimers: true,
        });
        await agent.start();

        // Three additions, no deletions.
        expect(storeSetSpy).toBeCalledTimes(3);
        expect(storeDeleteSpy).toBeCalledTimes(0);

        // Only the initial start-up check should have happened by now.
        expect(agent._timesChecked).toBe(1);

        // If it's one millisecond short of our next poll interval, nothing
        // should have changed yet.
        jest.advanceTimersByTime(defaultTestAgentPollSeconds * 1000 - 1);

        // Still only one check.
        expect(agent._timesChecked).toBe(1);

        // Now, we'll expect another request to go out, so we'll nock it.
        nockStorageSecret(genericServiceID, genericApiKeyHash);
        nockStorageSecretOperationManifest(
          genericServiceID,
          genericStorageSecret,
          304,
        );

        // If we move forward the last remaining millisecond, we should trigger
        // and end up with a successful sync.
        jest.advanceTimersByTime(1);

        // While that timer will fire, it will do work async, and we need to
        // wait on that work itself.
        await agent.requestPending();

        // Now the times checked should have gone up, and we should log that
        // we are looking for an update.  Of course, since we're unchanged
        // there should be no actual update.
        expect(agent._timesChecked).toBe(2);

        expect(storeSetSpy).toBeCalledTimes(3);
        expect(storeDeleteSpy).toBeCalledTimes(0);
      });

      it('purges operations which are removed from the manifest', async () => {
        const store = defaultStore();
        const storeSetSpy = jest.spyOn(store, 'set');
        const storeDeleteSpy = jest.spyOn(store, 'delete');

        // Intentionally not calling start, since we're not testing intervals.
        const agent = createAgent({ store });
        expect(storeSetSpy).toBeCalledTimes(0);

        nockStorageSecret(genericServiceID, genericApiKeyHash);
        nockGoodManifestsUnderStorageSecret(
          genericServiceID,
          genericStorageSecret,
          [
            sampleManifestRecords.a,
            sampleManifestRecords.b,
            sampleManifestRecords.c,
          ],
        ); // Starting with ABC.
        await agent.checkForUpdate();
        expect(agent._timesChecked).toBe(1);
        expect(storeSetSpy).toBeCalledTimes(3);
        expect(storeDeleteSpy).toBeCalledTimes(0);
        await expectStoreHasOperationEach(store, ['a', 'b', 'c']);

        nockStorageSecret(genericServiceID, genericApiKeyHash);
        nockGoodManifestsUnderStorageSecret(
          genericServiceID,
          genericStorageSecret,
          [sampleManifestRecords.a, sampleManifestRecords.b],
        ); // Just AB in this manifest.
        await agent.checkForUpdate();
        expect(agent._timesChecked).toBe(2);
        expect(storeSetSpy).toBeCalledTimes(3); // no new sets.
        expect(storeDeleteSpy).toBeCalledTimes(1);
        await expect(
          // Ensure that 'C' is gone!
          store.get(getStoreKey(sampleManifestRecords.c.signature)),
        ).resolves.toBeUndefined();

        nockStorageSecret(genericServiceID, genericApiKeyHash);
        nockGoodManifestsUnderStorageSecret(
          genericServiceID,
          genericStorageSecret,
          [sampleManifestRecords.a],
        ); // Just A in this manifest.
        await agent.checkForUpdate();
        expect(agent._timesChecked).toBe(3);
        expect(storeSetSpy).toBeCalledTimes(3); // no new sets.
        expect(storeDeleteSpy).toBeCalledTimes(2); // one more deletion
        await expect(
          // Ensure that 'B' is gone!
          store.get(getStoreKey(sampleManifestRecords.b.signature)),
        ).resolves.toBeUndefined();
      });

      describe('When given a graphVariant', () => {
        const graphVariant = 'different';
        const getOperationManifestRelativeUrl = (
          ...args: Parameters<typeof getOperationManifestUrl>
        ) =>
          getOperationManifestUrl(...args).replace(
            new RegExp(`^${urlOperationManifestBase}`),
            '',
          );

        it('fetches manifests for the corresponding variant', async () => {
          nockStorageSecret(genericServiceID, genericApiKeyHash);
          const agent = createAgent({
            apollo: {
              ...defaultAgentOptions.apollo,
              graphRef: `${genericServiceID}@${graphVariant}`,
            },
          });
          const nockedManifest = nockBase()
            .get(
              getOperationManifestRelativeUrl(
                genericServiceID,
                genericStorageSecret,
                graphVariant,
              ),
            )
            .reply(200, manifest(sampleManifestRecords.a));
          await agent.checkForUpdate();
          expect(nockedManifest.isDone()).toBe(true);
        });
      });
    });
  });
});


src/__tests__/common.test.ts
import * as common from '../common';

describe('common', () => {
  it('uses the correct cache prefix', () => {
    expect(common.getStoreKey('abc123')).toStrictEqual('abc123');
  });
  
  describe('urlOperationManifestBase', () => {
     const OLD_ENV = process.env;
     beforeEach(() => {
       jest.resetModules();
       process.env = { ...OLD_ENV };
     });
     afterAll(() => {
       process.env = OLD_ENV;
     });

     it('evaluates to the override url when overridden', () => {
       process.env.APOLLO_OPERATION_MANIFEST_BASE_URL = 'override';
       const { urlOperationManifestBase } = require('../common');
       expect(urlOperationManifestBase).toStrictEqual('override');
     });

     it('removes trailing slashes from the override url when overridden', () => {
       process.env.APOLLO_OPERATION_MANIFEST_BASE_URL = 'override/';
       const { urlOperationManifestBase } = require('../common');
       expect(urlOperationManifestBase).toStrictEqual('override');
     });

     it('evaluates to the test url when testing', () => {
       process.env.__APOLLO_OPERATION_REGISTRY_TESTS__ = 'true';
       const {
         urlOperationManifestBase,
         fakeTestBaseUrl,
       } = require('../common');
       expect(urlOperationManifestBase).toStrictEqual(fakeTestBaseUrl);
     });

     it('evaluates to the default value when not overridden or testing', () => {
       process.env.__APOLLO_OPERATION_REGISTRY_TESTS__ = 'false';
       const { urlOperationManifestBase } = require('../common');
       expect(urlOperationManifestBase).toStrictEqual(
         'https://operations.api.apollographql.com',
       );
     });
   });

   describe('urlStorageSecretBase', () => {
     const OLD_ENV = process.env;
     beforeEach(() => {
       jest.resetModules();
       process.env = { ...OLD_ENV };
     });
     afterAll(() => {
       process.env = OLD_ENV;
     });

     it('evaluates to the override url when overridden', () => {
       process.env.APOLLO_STORAGE_SECRET_BASE_URL = 'override';
       const { urlStorageSecretBase } = require('../common');
       expect(urlStorageSecretBase).toStrictEqual('override');
     });

     it('removes trailing slashes from the override url when overridden', () => {
       process.env.APOLLO_STORAGE_SECRET_BASE_URL = 'override/';
       const { urlStorageSecretBase } = require('../common');
       expect(urlStorageSecretBase).toStrictEqual('override');
     });

     it('evaluates to the test url when testing', () => {
       process.env.__APOLLO_OPERATION_REGISTRY_TESTS__ = 'true';
       const { urlStorageSecretBase, fakeTestBaseUrl } = require('../common');
       expect(urlStorageSecretBase).toStrictEqual(fakeTestBaseUrl);
     });

     it('evaluates to the default value when not overridden or testing', () => {
       process.env.__APOLLO_OPERATION_REGISTRY_TESTS__ = 'false';
       const { urlStorageSecretBase } = require('../common');
       expect(urlStorageSecretBase).toStrictEqual(
         'https://storage-secrets.api.apollographql.com',
       );
     });
   });
});


src/__tests__/helpers.test-helpers.ts
import { createHash } from '@apollo/utils.createhash';
import type { AgentOptions } from '../agent';
import {
  getStorageSecretUrl,
  urlStorageSecretBase,
  getOperationManifestUrl,
  urlOperationManifestBase,
  fakeTestBaseUrl,
} from '../common';
import nock from 'nock';
import { InMemoryLRUCache } from 'apollo-server-caching';
import type {
  Operation,
  OperationManifest,
} from '../ApolloServerPluginOperationRegistry';

export const defaultStore = () => new InMemoryLRUCache();

export const genericStorageSecret = 'someStorageSecret';
export const genericServiceID = 'test-service';
export const genericApiKeyHash = 'some-api-key-hash123';
export const defaultTestAgentPollSeconds = 60;

export const defaultAgentOptions: AgentOptions = {
  apollo: {
    graphRef: genericServiceID,
    keyHash: genericApiKeyHash,
  },
  store: defaultStore(),
  pollSeconds: defaultTestAgentPollSeconds,
};

// Each nock is good for exactly one request!

export function nockGoodManifestsUnderStorageSecret(
  graphId: string,
  storageSecret: string,
  operations: Operation[],
): nock.Scope {
  return nockStorageSecretOperationManifest(graphId, storageSecret, 200, {
    version: 2,
    operations,
  });
}

export function getStorageSecretPath(graphId: string, apiKeyHash: string) {
  return getStorageSecretUrl(graphId, apiKeyHash).replace(
    new RegExp(`^${urlStorageSecretBase}`),
    '',
  );
}

export function nockStorageSecret(
  graphId: string,
  apiKeyHash: string,
  status = 200,
  body: string = JSON.stringify(genericStorageSecret),
) {
  // Strip off the host for testing purposes with `nock`.
  return nockBase()
    .get(getStorageSecretPath(graphId, apiKeyHash))
    .reply(status, body);
}

export function getOperationManifestPath(
  graphId: string,
  storageSecret: string,
): string {
  // Strip off the host for testing purposes with `nock`.
  return getOperationManifestUrl(graphId, storageSecret, 'current').replace(
    new RegExp(`^${urlOperationManifestBase}`),
    '',
  );
}

export function nockStorageSecretOperationManifest(
  graphId: string,
  storageSecret: string,
  status = 200,
  body?: OperationManifest,
) {
  return nockBase()
    .get(getOperationManifestPath(graphId, storageSecret))
    .reply(status, body);
}

export function nockBase() {
  return nock(fakeTestBaseUrl);
}

export function hashApiKey(apiKey: string): string {
  return createHash('sha512').update(apiKey).digest('hex');
}

export function hashedServiceId(serviceID: string): string {
  return createHash('sha512').update(serviceID).digest('hex');
}


src/agent.ts
import type { ApolloConfig, WithRequired } from '@apollo/server';
import type { Logger } from '@apollo/utils.logger';
import type { fetch, RequestInit, Response } from 'apollo-server-env';
import type Keyv from 'keyv';
import loglevel from 'loglevel';
import fetcher from 'make-fetch-happen';
import type { OperationManifest } from './ApolloServerPluginOperationRegistry';
import { HttpRequestCache } from './cache';
import {
  getOperationManifestUrl,
  getStorageSecretUrl,
  getStoreKey,
  pluginName,
} from './common';

const DEFAULT_POLL_SECONDS: number = 30;
const SYNC_WARN_TIME_SECONDS: number = 60;

export interface AgentOptions {
  logger?: Logger;
  fetcher?: typeof fetch;
  pollSeconds?: number;
  apollo: WithRequired<ApolloConfig, 'keyHash' | 'graphRef'>;
  store: Keyv<string>;
}

type SignatureStore = Set<string>;

const callToAction = `Ensure this server's schema has been published with 'apollo service:push' and that operations have been registered with 'apollo client:push'.`;

export default class Agent {
  private fetcher: typeof fetch;
  private timer?: NodeJS.Timer;
  private logger: Logger;
  private requestInFlight: Promise<any> | null = null;
  private lastSuccessfulCheck?: Date;
  private storageSecret?: string;

  // Only exposed for testing.
  public _timesChecked: number = 0;

  private lastOperationSignatures: SignatureStore = new Set();
  private readonly options: AgentOptions = Object.create(null);

  // We've made most of our protocols capable of just taking a graph ref,
  // meaning that we can later iterate on the semantics of graph refs without
  // needing to make changes to Apollo Server. But not the operation registry!
  // We really need to know the pieces separately. *sigh*
  private readonly graphId: string;
  readonly graphVariant: string;

  constructor(options: AgentOptions) {
    Object.assign(this.options, options);

    const { graphRef } = this.options.apollo;
    const at = graphRef.indexOf('@');
    if (at === -1) {
      this.graphId = graphRef;
      this.graphVariant = 'current';
    } else {
      this.graphId = graphRef.substring(0, at);
      this.graphVariant = graphRef.substring(at + 1);
    }
    this.logger = this.options.logger || loglevel.getLogger(pluginName);
    this.fetcher = this.options.fetcher || getDefaultGcsFetcher();
  }

  async requestPending() {
    return this.requestInFlight;
  }

  private pollSeconds() {
    return this.options.pollSeconds || DEFAULT_POLL_SECONDS;
  }

  async start() {
    this.logger.debug('Starting operation registry agent...');

    // This is what we'll trigger at a regular interval.
    const pulse = async () => await this.checkForUpdate();

    // The first pulse should happen before we start the timer.
    try {
      await pulse();
    } catch (err) {
      console.error(
        'The operation manifest could not be fetched. Retries will continue, but requests will be forbidden until the manifest is fetched.',
        (err as Error).message || err,
      );
    }

    // Afterward, keep the pulse going.
    this.timer =
      this.timer ||
      setInterval(function () {
        // Errors in the interval indicate that the manifest might have failed
        // to update, but we've still got the seed update so we will continue
        // serving based on the previous manifest until we gain sync again.
        // These errors will be logged, but not crash the server.
        pulse().catch((err) => console.error(err.message || err));
      }, this.pollSeconds() * 1000);

    // Prevent the Node.js event loop from remaining active (and preventing,
    // e.g. process shutdown) by calling `unref` on the `Timeout`.  For more
    // information, see https://nodejs.org/api/timers.html#timers_timeout_unref.
    this.timer.unref();
  }

  public stop() {
    if (this.timer) {
      clearInterval(this.timer);
    }
  }

  private timeSinceLastSuccessfulCheck() {
    if (!this.lastSuccessfulCheck) {
      // So far back that it's never?
      return -Infinity;
    }
    return new Date().getTime() - this.lastSuccessfulCheck.getTime();
  }

  private warnWhenLossOfSync() {
    // This is probably good information to reveal in general, though nice
    // to have in development.
    if (this.timeSinceLastSuccessfulCheck() > SYNC_WARN_TIME_SECONDS * 1000) {
      console.warn(
        `WARNING: More than ${SYNC_WARN_TIME_SECONDS} seconds has elapsed since a successful fetch of the manifest. (Last success: ${this.lastSuccessfulCheck})`,
      );
    }
  }

  private async fetchAndUpdateStorageSecret(): Promise<string | undefined> {
    const storageSecretUrl = getStorageSecretUrl(
      this.graphId,
      this.options.apollo.keyHash,
    );

    const response = await this.fetcher(storageSecretUrl, this.fetchOptions);

    if (response.status === 304) {
      this.logger.debug(
        'The storage secret was the same as the previous attempt.',
      );
      return this.storageSecret;
    }

    if (!response.ok) {
      const responseText = await response.text();
      this.logger.debug(`Could not fetch storage secret ${responseText}`);
      return;
    }

    this.storageSecret = await response.json();

    return this.storageSecret;
  }

  private fetchOptions: RequestInit = {
    // More than three times our polling interval should be long enough to wait.
    timeout: this.pollSeconds() * 3 /* times */ * 1000 /* ms */,
  };

  private async fetchManifest(): Promise<Response> {
    this.logger.debug(`Checking for storageSecret`);
    const storageSecret = await this.fetchAndUpdateStorageSecret();

    if (!storageSecret) {
      throw new Error('No storage secret found');
    }

    const storageSecretManifestUrl = getOperationManifestUrl(
      this.graphId,
      storageSecret,
      this.graphVariant,
    );

    this.logger.debug(
      `Checking for manifest changes at ${storageSecretManifestUrl}`,
    );
    const response = await this.fetcher(
      storageSecretManifestUrl,
      this.fetchOptions,
    );

    if (response.status === 404 || response.status === 403) {
      throw new Error(
        `No manifest found for tag "${this.graphVariant}" at ` +
          `${storageSecretManifestUrl}. ${callToAction}`,
      );
    }
    return response;
  }

  private async tryUpdate(): Promise<boolean> {
    this._timesChecked++;

    let response: Response;
    try {
      response = await this.fetchManifest();
      // When the response indicates that the resource hasn't changed, there's
      // no need to do any other work.  Returning false is meant to indicate
      // that there wasn't an update, but there was a successful fetch.
      if (response.status === 304) {
        return false;
      }

      if (!response.ok) {
        const responseText = await response.text();
        // The response error code only comes in XML, but we don't have an XML
        // parser handy, so we'll just match the string.
        if (responseText.includes('<Code>AccessDenied</Code>')) {
          throw new Error(`No manifest found. ${callToAction}`);
        }
        // For other unknown errors.
        throw new Error(`Unexpected status: ${responseText}`);
      }

      const contentType = response.headers.get('content-type');
      if (contentType && contentType !== 'application/json') {
        throw new Error(`Unexpected 'Content-Type' header: ${contentType}`);
      }
    } catch (err) {
      const ourErrorPrefix = `Unable to fetch operation manifest for graph ID '${this.graphId}': ${err}`;

      (err as Error).message = `${ourErrorPrefix}: ${err}`;

      throw err;
    }

    await this.updateManifest(await response.json());
    // True is good!
    return true;
  }

  public async checkForUpdate() {
    // Display a warning message if things have fallen abnormally behind.
    this.warnWhenLossOfSync();

    // Don't check again if we're already in-flight.
    if (this.requestInFlight) {
      return this.requestInFlight;
    }

    // Prevent other requests from crossing paths.
    this.requestInFlight = this.tryUpdate();

    const resetRequestInFlight = () => (this.requestInFlight = null);

    return this.requestInFlight
      .then((result) => {
        // Mark this for reporting and monitoring reasons.
        this.lastSuccessfulCheck = new Date();
        resetRequestInFlight();
        return result;
      })
      .catch((err) => {
        // We don't want to handle any errors, but we do want to erase the
        // current Promise reference.
        resetRequestInFlight();
        throw err;
      });
  }

  public async updateManifest(manifest: OperationManifest) {
    if (
      !manifest ||
      manifest.version !== 2 ||
      !Array.isArray(manifest.operations)
    ) {
      throw new Error('Invalid manifest format.');
    }

    const incomingOperations: Map<string, string> = new Map();
    const replacementSignatures: SignatureStore = new Set();

    for (const { signature, document } of manifest.operations) {
      incomingOperations.set(signature, document);
      // Keep track of each operation in this manifest so we can store it
      // for comparison after the next fetch.
      replacementSignatures.add(signature);

      // If it it's _not_ in the last fetch, we know it's added.  We could
      // just set it — which would be less costly, but it's nice to have this
      // for debugging.
      if (!this.lastOperationSignatures.has(signature)) {
        // Newly added operation.
        this.logger.debug(`Incoming manifest ADDs: ${signature}`);
        this.options.store.set(getStoreKey(signature), document);
      }
    }

    // Explicitly purge items which have been removed since our last
    // successful fetch of the manifest.
    for (const signature of this.lastOperationSignatures) {
      if (!incomingOperations.has(signature)) {
        // Remove operations which are no longer present.
        this.logger.debug(`Incoming manifest REMOVEs: ${signature}`);
        this.options.store.delete(getStoreKey(signature));
      }
    }

    // Save the ones from this fetch, so we know what to remove on the next
    // actual update.  Particularly important since a future distributed
    // store might not actually let us look this up again.
    this.lastOperationSignatures = replacementSignatures;
  }
}

const GCS_RETRY_COUNT = 5;

function getDefaultGcsFetcher() {
  return fetcher.defaults({
    cacheManager: new HttpRequestCache(),
    // All headers should be lower-cased here, as `make-fetch-happen`
    // treats differently cased headers as unique (unlike the `Headers` object).
    // @see: https://git.io/JvRUa
    headers: {
      'user-agent': [
        require('../package.json').name,
        require('../package.json').version,
      ].join('/'),
    },
    retry: {
      retries: GCS_RETRY_COUNT,
      // The default factor: expected attempts at 0, 1, 3, 7, 15, and 31 seconds elapsed
      factor: 2,
      // 1 second
      minTimeout: 1000,
      randomize: true,
    },
  });
}
